### PR TITLE
removed a bunch of dead code. we may need it later or may not.

### PR DESCRIPTION
**Pull request by akiezun**
it's easy to bring it back if needed.

@lbergelson please review

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 1]
```diff
@@ -217,17 +216,6 @@ public SAMFileHeader getHeaderForReads() {
...
 
     /**
-     * Returns the header for the specified source of Features
-     *
-     * @param featureDescriptor FeatureInput whose header to retrieve
-     * @param <T> type of Feature in our FeatureInput
-     * @return header for the provided FeatureInput (null if we have no sources of Features)
-     */
-    public <T extends Feature> Object getHeaderForFeatures( final FeatureInput<T> featureDescriptor ) {
-        return hasFeatures() ? features.getHeader(featureDescriptor) : null;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: This is part of the tool API and should not be deleted (even if it's not currently used).

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java [hunk 2]
```diff
@@ -1,19 +0,0 @@
...
-import java.util.List;
-
-/**
- * An argument collection for use with tools that accept one or more input files containing reads
- * (eg., BAM/SAM/CRAM files), and require at least one such input.
- */
-public class RequiredReadInputArgumentCollection implements ArgumentCollectionDefinition {
-
-    @Argument(fullName = StandardArgumentDefinitions.INPUT_LONG_NAME, shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME, doc = "One or more BAM/SAM/CRAM files containing reads", optional = false)
-    public List<File> readFiles;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java]: These argument collections are part of the solution to the required/optional input issue addressed in https://github.com/broadinstitute/hellbender/pull/440/files and should not be deleted.

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java [hunk 3]
```diff
@@ -70,14 +70,6 @@ public Locatable getLocation() {
         return this.position;
     }
 
-    public List<String> getAllValues() {
-        return getValuesTo(values.size());
-    }
-
-    public List<String> getValuesTo(int columnPosition) {
-        return values.subList(0, columnPosition);
-    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java]: We agreed during the original codec code review that these methods were essential for making `TableFeature` actually usable by clients and should not be deleted (even if currently unused).

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: hmm, there were no tests for it though.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java]: there were no tests for it.

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java]: there were no tests

**Comment by akiezun** [comment]
fine. i'll bring those back for now.
i'll enter a ticket to write tests/examples for them.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/RequiredReadInputArgumentCollection.java]: ticket https://github.com/broadinstitute/hellbender/issues/451

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: ticket #452

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/table/TableFeature.java]: ticket #453

**Comment by akiezun** [comment]
back to @droazen

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalFeatureInputArgumentCollection.java [hunk 4]
```diff
@@ -19,4 +19,4 @@
     @Argument(fullName = StandardArgumentDefinitions.FEATURE_LONG_NAME, shortName = StandardArgumentDefinitions.FEATURE_SHORT_NAME, doc = "One or more files containing features", optional = true)
     public List<FeatureInput<Feature>> featureFiles;
 
-}
+}
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalFeatureInputArgumentCollection.java]: Revert these whitespace errors introduced into the ArgumentCollections

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 5]
```diff
@@ -218,13 +218,12 @@ public SAMFileHeader getHeaderForReads() {
...
     /**
      * Returns the header for the specified source of Features
-     *
      * @param featureDescriptor FeatureInput whose header to retrieve
      * @param <T> type of Feature in our FeatureInput
      * @return header for the provided FeatureInput (null if we have no sources of Features)
      */
     public <T extends Feature> Object getHeaderForFeatures( final FeatureInput<T> featureDescriptor ) {
-        return hasFeatures() ? features.getHeader(featureDescriptor) : null;
+         return hasFeatures() ? features.getHeader(featureDescriptor) : null;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: revert whitespace error

File: src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java [hunk 6]
```diff
@@ -112,18 +110,6 @@ public CannotHandleGzippedRef() {
...
-            super(String.format("Couldn't write file %s because %s with exception %s", file.getAbsolutePath(), message, getMessage(e)));
-        }
-
-        public CouldNotCreateOutputFile(File file, String message) {
-            super(String.format("Couldn't write file %s because %s", file.getAbsolutePath(), message));
-        }
-
-        public CouldNotCreateOutputFile(String filename, String message, Exception e) {
-            super(String.format("Couldn't write file %s because %s with exception %s", filename, message, getMessage(e)));
-        }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java]: These overloads will be needed as we port tools -- there's no good reason to delete them.

File: src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java [hunk 7]
```diff
@@ -78,23 +61,6 @@ public void setLane(final int lane) {
...
-     */
-    public void setOrCheckLane(final int lane) {
-        if (laneIsSet()) {
-            if (this.lane != lane) {
-                throw new IlluminaParserException("Lane number mismatch for " + this + " : " + this.lane + " != " + lane);
-            }
-        } else {
-            this.lane = lane;
-        }
-    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java]: Don't really like how you're poking random holes in a public API like this just for the sake of removing currently-unused code. You should be asking whether the API still make sense, and is actually usable and coherent with these holes poked into it...

File: src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ReadData.java [hunk 8]
```diff
@@ -52,17 +41,6 @@ public void setQualities(final byte[] qualities) {
...
-    /**
-     * @return Raw intensity values as produced by Illumina software, converted to shorts.
-     */
-    public FourChannelIntensityData getRawIntensities() {
-        return rawIntensities;
-    }
-
-    public void setRawIntensities(final FourChannelIntensityData rawIntensities) {
-        this.rawIntensities = rawIntensities;
-    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ReadData.java]: Again, removing random getters/setters like this just makes these classes less usable and consistent.

Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ReadData.java]: this class is going away anyway.
but these are not random getters. This is dead code. Dead code makes classes less usable and consistent.

File: src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java]: dead code is inherently evil.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java]: Inconsistent APIs are inherently evil -- you've harmed the semantics of these classes for no good reason.

**Comment by droazen** [comment]
Most of the remaining deletions seem reasonable -- I just object to the places where you've carved holes in standard, run-of-the-mill getter/setter type APIs. Consistency should trump desire to eliminate currently-unused methods in those few cases, I think.

File: src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/picard/illumina/parser/ClusterData.java]: these methods were dead - they have no semantics.
anyway, moot point. we're deleting this class.

File: src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalFeatureInputArgumentCollection.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/OptionalFeatureInputArgumentCollection.java]: done for all

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: done

File: src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java]: fine.

**Comment by akiezun** [comment]
i fixed what was requested. the illumina classes are scheduled for deletion (in another pull req)

**Pull request merged**